### DISCUSSION TITLE
Where is returns_rows set?

### DISCUSSION THREAD

**Discussion by fdcastel** [sqlalchemy/sqlalchemy#9759] [Usage Questions]
Working on Firebird dialect, I have [this test](https://github.com/sqlalchemy/sqlalchemy/blob/946e71efdfc93777027f4fd7360a524051be393d/lib/sqlalchemy/testing/suite/test_update_delete.py#L42) failing because `returns_rows` is `True` when it should not. 

Looking through SQLAlchemy code I can't find where it is set. 

What I found until now is that `_NoResultMetaData` has it set to `False`.

So I presume that in some part of my dialect I should have a way to inform if the result would be created as `_NoResultMetaData` or other `ResultMetaData`.

But these are all assumptions. Since I can't find any other Dialects doing the same, I'm kind of lost here. Please advice.

**Comment by zzzeek** [comment]
.returns_rows is set to True or False based on the presence of the ``cursor.description`` attribute on the DBAPI cursor.  Its value on CursorResult as you note is determined by the "self._metadata" object, which is normally expected to be either an instance of CursorResultMetaData (returns_rows=True) or _NoResultMetaData (returns_rows=False).    So it's the assignment of "self._metadata" in CursorResult that determines what this attribute will resolve as.  This occurs right in the constructor for `CursorResult` which you can see occurring for _NoResultMetaData here: https://github.com/sqlalchemy/sqlalchemy/blob/cef8adac8ae48c5c6c36138ed0b59067c1ef78ed/lib/sqlalchemy/engine/cursor.py#L1484   based on the conditional here:  https://github.com/sqlalchemy/sqlalchemy/blob/cef8adac8ae48c5c6c36138ed0b59067c1ef78ed/lib/sqlalchemy/engine/cursor.py#L1439   the "cursor_description" attribute is sent from DefaultDialect here: https://github.com/sqlalchemy/sqlalchemy/blob/cef8adac8ae48c5c6c36138ed0b59067c1ef78ed/lib/sqlalchemy/engine/default.py#L1793 or for a DML statement it's set here: https://github.com/sqlalchemy/sqlalchemy/blob/cef8adac8ae48c5c6c36138ed0b59067c1ef78ed/lib/sqlalchemy/engine/default.py#L1879 

Overall, dialects typically have no role in setting up whether or not cursor.description is present - this attribute needs to be non-None if the result has rows to be fetched, and needs to be None if there are not.     The only way that dialects might affect this is by using an alternate `CursorFetchStrategy`, which is usually the `BufferedRowCursorFetchStrategy`; this is used when the dialect is assembling a result set from data that was acquired somewhere other than the cursor.    

the test here seems to be a simple UPDATE statement without RETURNING, so there should be no cursor.description present.

**Reply by fdcastel** [reply]
Perfect, @zzzeek. Much appreciated. 

I was suspecting that the root cause of this problem could be in `firebird-driver`. 

With your detailed explanations I can now investigate further. 👍

**Reply by fdcastel** [reply]
Phew. That was fast. 😄 

```python
r.cursor
<firebird.driver.core.Cursor object at 0x000001F163266FB0>
r.cursor.description
()
special variables
function variables
len():
0
```

`r.cursor.description` is returning `()` instead of `None`.

I will talk with `firebird-driver` author about this.

Meanwhile, do you know what the DBAPI specification says about this? It should always be `None`? It could be an empty tuple?

**Reply by CaselIT** [reply]
> Meanwhile, do you know what the DBAPI specification says about this? It should always be `None`? It could be an empty tuple?

that seems a bug in the driver indeed. The spec says it needs to be None when not returning row: https://peps.python.org/pep-0249/#description

**Reply by CaselIT** [reply]
You may be able to workaround that until the dialect fixes the issue by setting up a custom `post_exec` https://docs.sqlalchemy.org/en/20/core/internals.html#sqlalchemy.engine.ExecutionContext.post_exec

**Reply by fdcastel** [reply]
Great!!! Thanks @CaselIT!

**Reply by fdcastel** [reply]
Sorry @CaselIT could you better explain what you had in mind?

My first (crude) attempt failed miserably:

```python

    def post_exec(self):    
        # Workaround for firebird-driver returning () instead of None.
        #   https://github.com/sqlalchemy/sqlalchemy/discussions/9759
        if self.cursor.description == ():
            self.cursor.description = None
```

Causes 
```
AttributeError("can't set attribute 'description'")
```

**Reply by CaselIT** [reply]
I guess the cursor object returned by the driver is immutable.

Try to see if using `object.__setattr__(self.cursor, 'description', None)` has more success

**Reply by fdcastel** [reply]
Unfortunately not. Same error.




But, now I see, `.description()` is a property: 

https://github.com/FirebirdSQL/python3-driver/blob/master/src/firebird/driver/core.py#L3928

Thanks again. I'll wait for `firebird-driver` author.

**Reply by zzzeek** [reply]
is this a new DBAPI for firebird?    the fdb and kinterbasdb drivers used to work ....

**Reply by CaselIT** [reply]
> Unfortunately not. Same error.
> 
> But, now I see, `.description()` is a property:
> 
> https://github.com/FirebirdSQL/python3-driver/blob/master/src/firebird/driver/core.py#L3928
> 
> Thanks again. I'll wait for `firebird-driver` author.

It seem you can replace `self._stmt._desc = None`. But it probably not very robust

**Reply by fdcastel** [reply]
> is this a new DBAPI for firebird? the fdb and kinterbasdb drivers used to work ....

Yes. A modern driver using the new Firebird API. For Firebird 3.0+

`fdb` will be mantained for Firebird 2.5 compatibility. `kinterbasdb` is being removed.

**Comment by fdcastel** [comment]
All done!

I submitted [a PR](https://github.com/FirebirdSQL/python3-driver/pull/21) yesterday and it was merged.

**Reply by CaselIT** [reply]
ok great!
